### PR TITLE
Add native manual reference examples

### LINKED ISSUE #1: Release artifacts (PaperBanana framework code, PaperBananaBench dataset) on Hugging Face
Hello,

Niels here from the open-source team at Hugging Face. I discovered your work through Hugging Face's daily papers as yours got featured: https://huggingface.co/papers/2601.23265.
The paper page lets people discuss about your paper and lets them find artifacts about it (your models, datasets or demo for instance), you can also claim
the paper as yours which will show up on your public profile at HF, add Github and project page URLs.

I saw on your GitHub repository (https://github.com/dwzhu-pku/PaperBanana) that your code and the PaperBananaBench dataset will be released in ~2 weeks. That's fantastic news!
It'd be great to make the code for your PaperBanana framework and the PaperBananaBench dataset available on the 🤗 hub, to improve their discoverability/visibility once they are released.
We can add tags so that people find them when filtering https://huggingface.co/models and https://huggingface.co/datasets.

## Uploading models

See here for a guide: https://huggingface.co/docs/hub/models-uploading.

In this case, we could leverage the [PyTorchModelHubMixin](https://huggingface.co/docs/huggingface_hub/package_reference/mixins#huggingface_hub.PyTorchModelHubMixin) class which adds `from_pretrained` and `push_to_hub` to any custom `nn.Module`. Alternatively, one can leverages the [hf_hub_download](https://huggingface.co/docs/huggingface_hub/en/guides/download#download-a-single-file) one-liner to download a checkpoint from the hub.

We encourage researchers to push each model checkpoint to a separate model repository, so that things like download stats also work. We can then also link the checkpoints to the paper page.

## Uploading dataset

Would be awesome to make the dataset available on 🤗 , so that people can do:

```python
from datasets import load_dataset

dataset = load_dataset("your-hf-org-or-username/your-dataset")
```
See here for a guide: https://huggingface.co/docs/datasets/loading.

Besides that, there's the [dataset viewer](https://huggingface.co/docs/hub/en/datasets-viewer) which allows people to quickly explore the first few rows of the data in the browser.

Let me know if you're interested/need any help regarding this once they are released!

Cheers,

Niels
ML Engineer @ HF 🤗

### LINKED ISSUE #10: 是否会开放网页版试用Demo？

### LINKED ISSUE #24: Project get suspended by Google
my project is being suspended for repeatedly violating our[ Google Cloud Platform Terms of Service](https://cloud.google.com/terms/) or[ Acceptable Use Policy](https://cloud.google.com/terms/aup) (including[ Terms of Service](https://console.developers.google.com/terms) of the Google API you may be using).

does anyone encounter the similar issue?

### LINKED ISSUE #32: Would you be open to a visual storytelling enhancement for the Planner agent?
Hi! First off — thank you for building PaperBanana. I've been using it extensively and the multi-agent pipeline architecture is genuinely impressive.

While working with it, I experimented with adding a **visual metaphor discovery step** to the Planner agent — before describing boxes and arrows, it first asks "What is this concept LIKE in the real world?" and finds a visual analogy. The metaphor then guides the entire diagram.

I tested across 4 scenarios using the same image generation model (only changing what the Planner asks the model to draw):

| Scenario | With Metaphor | Baseline | Delta |
|----------|:---:|:---:|:---:|
| Application ecosystem | 93 | 68 | +25 |
| Product overview | 94 | 78 | +16 |
| Technical architecture | 95 | 65 | +30 |
| Abstract concepts | 92 | 76 | +16 |
| **Average** | **93.5** | **71.75** | **+21.75** |

I also built an SVG rendering mode (Cairo-based, 100% text rendering fidelity) and a vision critic loop, but I understand that's a bigger change.

**I submitted PR #30 but saw it was closed** — totally understand, and no hard feelings at all. I realize I should have opened a discussion first before submitting a large PR. That's on me.

So I wanted to ask: **would you be interested in any of these enhancements?** I'm happy to:

- Submit a small, focused PR with just the Planner metaphor step (minimal diff)
- Split the work into separate small PRs (one per agent change)
- Provide the changes in whatever format is easiest for your team to review
- Or just keep it as a community fork if you'd prefer — no pressure at all

My fork is at [stuinfla/paperbanana](https://github.com/stuinfla/paperbanana) if you want to look at the results first. The [showcase gallery](https://stuinfla.github.io/paperbanana/) has 10 example diagrams.

Thanks for your time, and thanks again for making PaperBanana open source. It's a great project.

— Stuart ([@stuinfla](https://github.com/stuinfla))

### LINKED ISSUE #45: Is there a model that is more suitable for this project?
I see that this project currently uses Google's API, so is there a more suitable model for this project, and if you use the on-premises model, how should I choose, I hope to get your suggestions.

### LINKED ISSUE #53: Possibility to use Ollama backend instead of using API keys?
If someone has decent compute hardware on-prem, is it possible to use Ollama or other alternatives instead of using paid API keys?

### LINKED ISSUE #54: Bug: Figure Size setting in Gradio UI has no effect on generated images
## Description

The "Figure Size" dropdown in the Gradio web UI (`app.py`) accepts user input but the selected value is never passed to the image generation pipeline. All generated images use a hardcoded `image_size="1k"` regardless of the Figure Size selection.

## Steps to Reproduce

1. Open the Gradio UI (`python app.py`)
2. In Settings, change "Figure Size" to any value (e.g., "14-17cm")
3. Generate candidates
4. Observe that the output image resolution is always 1k

## Root Cause

In `app.py`, `run_generate()` receives `figure_size` and stores it in a local dict:

```python
params = {"figure_size": figure_size}
```

But `params` is never passed to `process_parallel_candidates()` or any downstream function. The variable is effectively dead code.

Meanwhile, in `agents/visualizer_agent.py`, `image_size` is hardcoded to `"1k"` in both the Gemini and OpenRouter image generation paths (lines 192 and 206):

```python
"image_size": "1k",  # hardcoded, ignores UI setting
```

## Suggested Fix

1. Pass `figure_size` through `additional_info` in `create_sample_inputs()`
2. In `VisualizerAgent`, read `figure_size` from `data["additional_info"]` and map it to an appropriate `image_size` value:

| Figure Size | Suggested image_size |
|---|---|
| 1-3cm, 4-6cm | 1k |
| 7-9cm, 10-13cm | 2k |
| 14-17cm | 4k |

## Environment

- PaperBanana version: `fed1812` (latest main)
- Gradio UI (`app.py`)

### LINKED ISSUE #59: You exceeded your current quota, please check your plan and billing details
You exceeded your current quota, please check your plan and billing details
How to solve this situation？

### LINKED ISSUE #61: 疑似HuggingFace Demo有跨会话 API key 复用风险
## 问题描述
在填入API后，没有任何操作，API后台显示有大量消耗，几分钟内消耗了13刀。怀疑 #60 也是同样的原因。
## 可能的问题原因
在app.py中，apply_keys()函数会将用户输入写入：
os.environ["OPENROUTER_API_KEY"]
os.environ["GOOGLE_API_KEY"]
随后调用reinitialize_clients()。
在utils/generation_utils.py中，reinitialize_clients()会重建模块级全局客户端，例如：
gemini_client
openrouter_client
后续run_generate()与run_refine()并未接收会话级 API 密钥作为输入。它们调用下游生成 / 优化逻辑时，会使用当前已初始化的任意全局客户端。
服务路由模块同样会自动选择当前可用的全局客户端。

这意味着某一浏览器会话中的用户操作，可能影响由同一运行中的 Space 进程处理的、后续来自无关会话的请求。
用户 A 输入 API 密钥并点击 “应用密钥”
Space 进程将该密钥存入进程全局状态
用户 B 在另一会话中打开该 Space 并提交请求，未提供自身密钥
请求仍可能通过 A 此前初始化的客户端执行

### LINKED ISSUE #64: Critic agent should optionally read the style guide to avoid degrading aesthetics across rounds
## Observation

When running the `demo_full` (or `demo_planner_critic`) pipeline with multiple Critic rounds, the **first-round image is consistently the most visually coherent**, and successive Critic rounds **progressively degrade the aesthetic** — re-rendered images become more cluttered, lose colour discipline, and drift away from the Stylist's choices.

Setting `--max_critic_rounds=0` (or 1) and skipping the Critic gives the best output for many use cases. This suggests the issue isn't the Critic per se but a missing input to it.

## Diagnosis

The Stylist agent (`agents/stylist_agent.py`) and Polish agent (`agents/polish_agent.py`) both load `style_guides/{prefix}_{task}_style_guide.md` and feed it to their model calls. The Critic agent (`agents/critic_agent.py`) does **not** — it only sees the raw content + caption + image and is prompted to identify fidelity gaps ("which named components are missing or wrong in the rendered figure?").

This means:

1. The Stylist makes deliberate aesthetic choices guided by the style guide (palette, typography, shape language, whitespace, illustration register).
2. The Critic, blind to those choices, flags every named-but-not-prominent component as a fidelity violation and asks the Visualizer to add it back.
3. The Visualizer's revision is applied destructively and erodes the aesthetic the Stylist set up.
4. After 2–3 rounds the figure is "more faithful" by the Critic's metric but visibly worse by the Stylist's.

## Reproducible test

1. Generate a figure in `demo_full` mode with `max_critic_rounds=3`.
2. Inspect the per-round images saved to disk: `target_{task}_stylist_desc0_base64_jpg` (round 0, post-Stylist) vs. `target_{task}_critic_desc0_base64_jpg`, `target_{task}_critic_desc1_base64_jpg`, etc.
3. The post-Stylist image is typically cleaner and more on-brand than the post-Critic ones, especially for icon-rich diagrams or editorial-style illustrations where label density should be intentionally low.

## Proposed fix

Have the Critic load the same style guide the Stylist did and prepend it to the system prompt. The Critic's job description shifts from "find fidelity gaps" to "find fidelity gaps that don't break the style guide" — its suggestions reinforce the Stylist's choices instead of fighting them.

Concrete sketch (mirrors the existing pattern in `agents/stylist_agent.py:62`):

```python
# agents/critic_agent.py
import os
...
class CriticAgent(BaseAgent):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.model_name = self.exp_config.main_model_name

        style_prefix = os.environ.get("PAPERBANANA_STYLE_GUIDE_PREFIX", "neurips2025")
        task = "plot" if self.exp_config.task_name == "plot" else "diagram"
        guide_path = self.exp_config.work_dir / f"style_guides/{style_prefix}_{task}_style_guide.md"
        try:
            with open(guide_path, "r", encoding="utf-8") as f:
                self.style_guide = f.read()
        except FileNotFoundError:
            self.style_guide = ""
        ...
```

Then include `self.style_guide` in the user prompt the Critic sends to the model, with instructions like:

> Before suggesting changes, verify each suggestion would not violate the style guide below. If a fidelity gap can only be closed by violating the style guide (e.g., by adding more text labels than the guide allows), prefer to leave the figure as-is.

## Why this is worth doing upstream

- It's a small, contained change (one agent, ~30 lines) that doesn't break the existing API.
- The Critic-vs-Stylist conflict is structural — anyone using `demo_full` with non-default style guides will hit it.
- It removes a class of "the Critic made my figure worse" complaints that currently can only be worked around by lowering `max_critic_rounds` to 0–1.
- Optional: gate it behind a flag (`--critic-uses-style-guide`) for backwards compatibility.

## Adjacent observations

While digging into this, two other small things in the same area:

1. The Streamlit UI in `demo.py` has `min_value=1` on the Max Critic Rounds input, which prevents users from setting it to 0 even though the pipeline supports it (`for round_idx in range(max_rounds)` with `max_rounds=0` is a clean no-op). Lowering the min would give users an immediate workaround.
2. The Critic loop has no quality regression check — it only rolls back if the Visualizer fails to return an image (`utils/paperviz_processor.py:88-97`). It assumes monotonic improvement across rounds, which the above issue undermines. A "did the new image actually score better than the previous?" gate (using e.g. CLIP similarity to the Stylist's first render or a separate VLM eval) would let the loop self-correct.

Happy to send a PR for the style guide change if helpful — let me know if the maintainers would prefer a particular approach (always-on vs. flag-gated, prompt wording, etc.) before I start.

### PR DESCRIPTION
**Pull request by jdotc1**
## Summary
- add a native PaperBananaBench reference loader and compact Prompt Studio selector for diagram and plot examples
- enrich provider prompts with up to 10 manually selected examples without mutating the editor prompt
- persist selected reference provenance in request.json, generated metadata, and provider request prompt payloads
- keep selections benchmark-scoped so diagram requests keep only diagram references and plot requests keep only plot references
- surface missing local benchmark images as non-blocking selector warnings with row-level image status
- decode and surface selected-reference provenance in Run Details, Run Ledger, Artifact Library, search, and assistant context
- extend legacy Gradio, Streamlit, skill, and visualization surfaces to carry task_name through diagram and plot workflows
- share legacy result-key resolution across UI surfaces so stale cross-task eval_image_field values are ignored, explicit Critic mode uses the latest critic round, and polished plot outputs are treated as final Auto outputs when present
- thread legacy Figure Size into provider image_size selection for visualizer, vanilla, and polish generation
- normalize multi-series plot JSON strings before the legacy planner/visualizer path sees them
- load task style guides in the Critic and allow zero critic rounds in legacy UI surfaces
- remove the Gradio cross-session API-key mutation path from the legacy surface and guard it with tests
- add an opt-in native Gemini code-execution Critic mode through --agentic-critic, preserving text/code/result evidence in intermediate records
- add an explicit local OpenAI-compatible text route for legacy Python model names prefixed with local/ or ollama/
- add FAQ/troubleshooting/contribution docs for Hugging Face artifacts, hosted demo availability, quota/suspension support, provider/model choice, API-key safety, Ollama/local limits, and large planner/prompt proposals
- update README/skill/UI copy so native manual references and legacy workflows no longer describe plots as unsupported

## Issue Closeout
- Fixes #1
- Fixes #10
- Fixes #24
- Fixes #32
- Fixes #35
- Fixes #45
- Fixes #53
- Fixes #54
- Fixes #59
- Fixes #61
- Fixes #64
- Fixes #67

## Validation
- ./script/check_xcode_project_drift.sh
- ./script/check_native_source_control_contract.sh
- git diff --cached --check
- PYTHONDONTWRITEBYTECODE=1 .venv/bin/python -m unittest tests.test_agentic_critic tests.test_local_openai_route
  - 9 tests passed
- PYTHONDONTWRITEBYTECODE=1 .venv/bin/python -m unittest tests.test_legacy_generation_options tests.test_legacy_plot_agents tests.test_legacy_plot_contracts tests.test_native_generate_cli tests.test_skill_run
  - 31 tests passed
- PYTHONDONTWRITEBYTECODE=1 .venv/bin/python -m pytest -q -p no:cacheprovider tests
  - 69 tests passed
- DEVELOPER_DIR=/Applications/Xcode-beta.app/Contents/Developer ./script/test_all.sh
  - source-control contract passed
  - Xcode project drift check passed
  - Xcode 27 baseline guard passed
  - Swift: 153 tests passed under the repeated xcodebuild test command
  - Python: 69 tests passed
  - Xcode 27 proof: status=passed halted=False
- DEVELOPER_DIR=/Applications/Xcode-beta.app/Contents/Developer ./script/build_and_run.sh --no-open
  - build succeeded

## Notes
- local PaperBananaBench data is not downloaded automatically by the app; the selector shows setup, missing, malformed, or empty states when data/PaperBananaBench/{diagram,plot}/ref.json is unavailable
- benchmark images are not copied into native run folders; durable artifacts store selected reference metadata and enrich only the provider prompt
- the local/Ollama route is intentionally text-only and explicit; it does not silently replace image-generation providers
- agentic Critic code execution is intentionally opt-in and Gemini-only for v1 so the default model router behavior remains unchanged
- support/discussion issues are handled by durable docs and contribution guidance rather than by automatic provider account changes or automatic dataset downloads

### CODE REVIEW

**Comment by jdotc1** [comment]
Closing this broad PR unmerged as superseded by the focused closeout series:

- #69 covers legacy plot routing and figure sizing for #54 and #67.
- #70 covers hosted UI credential isolation for #61.
- #71 covers the opt-in planner metaphor mode for #32.
- #72 carries the first-class native macOS app and native manual reference work from this PR.
- #73 covers critic controls and agentic critic mode for #35 and #64.
- #74 covers local/Ollama text routing plus provider/support documentation for #1, #10, #24, #45, #53, and #59.

This keeps review and merge risk bounded while avoiding #68 remaining as a duplicate issue-closing vehicle. The original integrated safety branch remains preserved as `jdotc1/integration-12-issue-closeout-35ee29f` at `35ee29f`.
